File: adventofcode/2021/day13/day13.rb
# frozen_string_literal: true

def read_input
  ARGF.read
end

def display(coords)
  xmax = coords.map(&:first).max + 1
  ymax = coords.map(&:last).max + 1
  cmap = coords.each_with_object({}) { |c, r| r[c] = true }
  ymax.times do |y|
    xmax.times do |x|
      print cmap[[x, y]] ? '#' : '.'
    end
    puts
  end
  puts
end

def fold_hor(coords, row)
  # fold up
  coords.map do |x, y|
    next [x, y] if y <= row

    [x, 2 * row - y]
  end
end

def fold_ver(coords, col)
  # fold left
  coords.map do |x, y|
    next [x, y] if x <= col

    [2 * col - x, y]
  end
end

def part1(coords, insts)
  c, val = insts[0] # only first
  coords = if c == 'x'
             fold_ver(coords, val.to_i)
           else
             fold_hor(coords, val.to_i)
           end
  coords.uniq.size
end

def part2(coords, insts)
  insts.each do |inst|
    c, val = inst
    coords = if c == 'x'
               fold_ver(coords, val.to_i)
             else
               fold_hor(coords, val.to_i)
             end
  end
  display coords
end

coords, insts = read_input.chomp.split("\n\n")
coords = coords.split("\n").map { |c| c.split(',').map(&:to_i) }
insts = insts.split("\n").map { |i| i.split(' ').last.split('=') }

p part1(coords.map(&:dup), insts)
part2(coords.map(&:dup), insts)

File: adventofcode/2021/day13/day13.hs
{-# LANGUAGE TupleSections #-}

import Data.List
import Debug.Trace (trace)
import qualified Data.Map as M

data Point = Point Int Int
  deriving (Eq, Ord, Show)

newtype Inst = Inst Point
  deriving Show

splitLines :: [String] -> String -> [[String]]
splitLines [] _ = []
splitLines lines pat
  | null rs = [ss]
  | otherwise = ss : splitLines rs pat
  where ss = takeWhile (/= pat) lines
        rs = dropWhile (== pat) $ dropWhile (/= pat) lines

splitStr :: Char -> String -> [String]
splitStr _ [] = []
splitStr ch cs
  | null rs = [ss]
  | otherwise = ss : splitStr ch rs
  where ss = takeWhile (/= ch) cs
        rs = dropWhile (== ch) $ dropWhile (/= ch) cs

parseCoords :: [String] -> [Point]
parseCoords cs = ps
  where f = splitStr ','
        ps = map ((\[x, y] -> Point x y) . (map read . f)) cs

parseInsts :: [String] -> [Inst]
parseInsts cs = ps
  where f = splitStr '='
        p :: [String] -> Point
        p ["x", n] = Point (read n) 0
        p ["y", n] = Point 0 (read n)
        p [_] = Point 0 0
        ps = map (Inst . p . f . last . words) cs

fld :: [Point] -> Inst -> [Point]
fld ps i = fs
  where (Inst (Point x y)) = i
        f :: Int -> Int -> Int
        f cx x
          | x == 0 = cx
          | otherwise = x - abs (x - cx)
        fs = map (\(Point cx cy) -> Point (f cx x) (f cy y)) ps

part1 :: [Point] -> Inst -> Int
part1 ps i = length $ nub $ fld ps i

display :: [Point] -> String
display ps = unlines $ map (map f) aps
  where m = M.fromList $ map (, 1) ps
        xmax = maximum $ map (\(Point x _) -> x) ps
        ymax = maximum $ map (\(Point _ y) -> y) ps
        aps = [[Point x y | x <- [0..xmax]] | y <- [0..ymax]]
        f :: Point -> Char
        f p
          | M.member p m = '#'
          | otherwise = ' '


part2 :: [Point] -> [Inst] -> String
part2 ps is = display fs
  where fs = foldl fld ps is


solve :: [String] -> String
solve ls = unlines [show p1, p2]
  where lines = splitLines ls ""
        coords = parseCoords $ head lines
        insts = parseInsts $ last lines
        p1 = part1 coords (head insts)
        p2 = part2 coords insts

main :: IO ()
main = interact $ solve . lines
